server/cluster/coordinator.go

```diff
@@ -820,3 +822,130 @@ func (s *scheduleController) IsPaused() bool {
+	stage                 UnsafeRecoveryStage
+	failedStores          map[uint64]bool
+	storeReports          map[uint64]*pdpb.StoreReport // Store info proto
+	numStoresReported     uint64
+	storeRecoveryPlans    map[uint64]*pdpb.RecoveryPlan // StoreRecoveryPlan proto
+	numStoresPlanExecuted uint64
+}
+
+func NewUnsafeRecoveryController(coordinator *coordinator) *UnsafeRecoveryController {
+	return &UnsafeRecoveryController{
+		coordinator:           coordinator,
+		stage:                 Ready,
+		failedStores:          make(map[uint64]bool),
+		storeReports:          make(map[uint64]*pdpb.StoreReport),
+		numStoresReported:     0,
+		storeRecoveryPlans:     make(map[uint64]*pdpb.RecoveryPlan),
+		numStoresPlanExecuted: 0,
+	}
+}
+
+func (u *UnsafeRecoveryController) RemoveFailedStores(failedStores map[uint64]bool) bool {
+	u.Lock()
+	defer u.Unlock()
+	if len(u.failedStores) != 0 {
+		return false
+	}
+	u.failedStores = failedStores
+	for _, s := range coordinator.rc.GetStores() {
+		if s.IsTombstone() || s.IsPhysicallyDestroyed() || failedStores[s.GetID()] {
+			continue
+		}
+		u.storeResponse[s.GetID()] = nil
+	}
+	u.stage = CollectingClusterInfo
+	return true
+}
+
+func (u *UnsafeRecoveryController) HandleStoreHeartbeat(heartbeat *pdpb.StoreHeartbeatRequest, resp *pdpb.StoreHeartbeatResponse) {
+	u.Lock()
+	defer u.Unlock()
+	if len(failedStores) == 0 {
+		return
+	}
+	switch u.stage {
+	case CollectingClusterInfo:
+		if heartbeat.DetailedReport.StoreID == 0 {
+			// Inform the store to send detailed report in the next heartbeat.
+			resp.ReportDetailsInNextHeartbeat = true
+		} else if u.storeReports[heartbeat.StoreReport.StoreID] == nil {
+			u.storeReports[heartbeat.StoreReport.StoreID] = heartbeat.StoreReport
+			u.numStoresResponded++
+			if u.numStoresResponded == len(storeReports) {
+				// Info collection is done.
+				u.stage = Recovering
+				go u.GenerateRecoveryPlan()
+			}
+		}
+	case Recovering:
+		if u.storeRecoveryPlans[heartbeat.StoreReport.StoreID] != nil {
+			if !u.IsPlanExecuted(heartbeat.StoreReport) {
+				// If the plan has not been executed, send it through the heartbeat response.
+				resp.RecoveryPlan = u.storeRecoveryPlans[heartbeat.StoreReport.StoreID]
+			} else {
+				u.numStoresPlanExecuted++
+				if u.numStoresPlanExecuted == len(storeRecoveryPlans) {
+					// The recovery is finished.
+					u.stage = Ready
+					u.failedStores = make(map[uint64]bool)
+					u.storeReports = make(map[uint64]*pdpb.StoreReport)
+					u.numStoresReported = 0
+					u.storeRecoveryPlans = make(map[uint64]*pdpb.RecoveryPlan)
+					u.numStoresPlanExecuted = 0
+				}
+			}
+
+		}
+	}
+}
+
+func (u *UnsafeRecoveryController) GenerateRecoveryPlan() {
+    u.Lock()
+    defer u.Unlock()
+}
+
+func (u *UnsafeRecoveryController) Show() string {
+	u.RLock()
+	defer u.RUnlock()
+	switch u.stage {
+	case Ready:
+		return "Ready"
+	case CollectingClusterInfo:
+		return "Collecting cluster info from all alive stores..."
```

Comment:
show the unfinished store ids and progress (x/n where x the count of finished stores and n is the total number of stores)